examples/ex4_0.ipynb

```diff
@@ -721,7 +721,7 @@
    "cell_type": "markdown",
    "metadata": {},
    "source": [
-    "In this sample the conditional probability `P(sez_noodle|has_butter)` is approximately 1:2, while `P(has_butter|sez_noodle)` is approximately 1:4 (i.e., significantly lower).\n",
+    "We added (has_butter, sez_noodle) tuples. In this sample the conditional probability `P(sez_noodle|has_butter)` is approximately 1:2, while `P(has_butter|sez_noodle)` is approximately 1:4 (i.e., significantly lower).\n",
```

Comment:
I'm second guessing myself on my conditional probas 😁 

```raw
| random variable |       | sez_noodle | sez_noodle |
|                 |       | False      | True       |
|-----------------|-------|------------|------------|
| has_butter      | False | 27         | 15         |
| has_butter      | True  | 53         | 7          |
```

P(sez_noodle|has_butter) = P(sez_noodle and has_butter) / P(has_butter)
= 7 / (7 + 53)
~= 0.12 ~= 1/8 = 1:7 odd

P(has_butter|sez_noodle) = P(has_butter and sez_noodle) / P(sez_noodle)
= 7 / (7 + 15)
~= 0.31 ~= 1/3 = 1:2 odd

Reply:
The edits are looking good!  And much appreciated :)

Also, I am completely embarrassed about my calculation mistakes there!! 🤦‍♂️ Thank you so much for catching that.

Using the `collections.Counter` in that notebooks felt kind of kludgy to me... and it's really not a *confusion matrix* at all.  I was looking for better ways to illustrate this part about conditional probabilities -- perhaps within a dataframe or visualization? Ultimately, integrating [`pgmpy`](https://pgmpy.org/) would be a big improvement :)  Do you have any suggestions about better ways to illustrate this part?

Reply:
Aha not embarrassing at all! I'm really enjoying reviewing the material that changed quite a bit since the KConnexions tuto. Lots of work went in 👏 . (btw, I'm still mind blown by https://derwen.ai/docs/kgl/overview/#how-to-use-these-materials 🤯 really cool to see how deliberate your approach is ... reminds me of https://documentation.divio.com/ but like a much better version)

Yeah I agree that visualisation is a big part of conditional probas. I didn't know `pgmpy`, will have a quick look.
In terms of suggestion, I would go either with 1) pd table or 2) mosaic plot.

1)
```python
df = pd.DataFrame(
    [[27, 15], [53, 7]], 
    columns=pd.MultiIndex.from_tuples([("sez_noodle", False), ("sez_noodle", True)]),
    index=pd.MultiIndex.from_tuples([("has_butter", False), ("has_butter", True)])
)
# add totals
df.loc['Total']= df.sum()
df['Total'] = df.sum(axis=1)
```
adding totals like this is kinda hacky and the result is this 
![image](https://user-images.githubusercontent.com/7823843/107490651-748d2c80-6b8a-11eb-8d86-734f155b0b1f.png)
this data is semantically a pd.Series in my opinion, with a MultiIndex of rank=2 so displaying it in this way is a bit misleading maybe

2) I used https://www.statsmodels.org/stable/generated/statsmodels.graphics.mosaicplot.mosaic.html
```python
from statsmodels.graphics.mosaicplot import mosaic
from matplotlib import pyplot as plt

contingency_table = {(False, False): 27, (False, True): 15, (True, False): 53, (True, True): 7}
fig, ax = plt.subplots(1,1)
fig, _ = mosaic(
    contingency_table,
    # warning: mosaic expects data in (x_axis, y_axis), so here, I revert the axis so that the contingency table has the same order
    index=[1, 0], 
    gap=0.05, 
    title='recipes',
    # default cmap isn't great, so I change it here
    properties=lambda key: {'color': 'yellow' if key[1] == 'True' else 'gray'},
    ax=ax
)
ax.set_xlabel('sez_noddle')  
ax.set_ylabel('has_butter')
ax.set_box_aspect(1)
```
![image](https://user-images.githubusercontent.com/7823843/107493997-896bbf00-6b8e-11eb-8731-d8793792b5fe.png)



Anyways, will continue to review the material today.

Reply:
Wonderful! Using a mosaic plot works really well, and `statsmodel` will likely get used elsewhere in `kglab` too.

Reply:
@louisguitton I've merged the contingency_table/mosaic into the tutorial, and added `statsmodel` as a dependency.

FWIW, my installation of `matplotlib` was causing the dreaded unfilterable "deprecation warnings" 🧐 which required much reworking of pip and conda installations on my Mac laptop.  I'm so glad to have received an evaluation device of https://kfocus.org/ to use -- less problems with conda and brew fighting over header files!